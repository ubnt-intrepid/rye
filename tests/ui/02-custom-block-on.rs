fn main() {}

fn custom_block_on<F>(fut: F) -> F::Output
where
    F: std::future::Future + Send + 'static,
{
    futures_executor::block_on(fut)
}

#[rye::test_case(block_on = "custom_block_on")]
async fn nested() {
    use futures_test::future::FutureTestExt;

    let mut vec = vec![0usize; 5];
    let rc = std::rc::Rc::new(());

    assert_eq!(vec.len(), 5);
    assert!(vec.capacity() >= 5);

    section!("resizing bigger changes size and capacity", {
        vec.resize(10, 0);

        async {}.pending_once().await;

        assert_eq!(vec.len(), 10);
        assert!(vec.capacity() >= 10);

        section!("shrinking smaller does not changes capacity", {
            vec.resize(0, 0);

            assert_eq!(vec.len(), 0);
            assert!(vec.capacity() >= 10);
        });
    });

    drop(rc);
}

#[rye::test_case(block_on = "futures_executor::block_on")]
async fn case_async() {
    let mut vec = vec![0usize; 5];

    assert_eq!(vec.len(), 5);
    assert!(vec.capacity() >= 5);

    section!("resizing bigger changes size and capacity", {
        vec.resize(10, 0);

        assert_eq!(vec.len(), 10);
        assert!(vec.capacity() >= 5);
    });
}
