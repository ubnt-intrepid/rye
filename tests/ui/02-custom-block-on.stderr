error: future cannot be sent between threads safely
  --> $DIR/02-custom-block-on.rs:10:29
   |
3  | fn custom_block_on<F>(fut: F) -> F::Output
   |    ---------------
4  | where
5  |     F: std::future::Future + Send + 'static,
   |                              ---- required by this bound in `custom_block_on`
...
10 | #[rye::test_case(block_on = "custom_block_on")]
   |                             ^^^^^^^^^^^^^^^^^ future returned by `nested` is not `Send`
   |
   = help: within `impl std::future::Future`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::RefCell<std::collections::HashMap<&'static rye::section::SectionId, rye::section::SectionData>>>`
note: future is not `Send` as this value is used across an await
  --> $DIR/02-custom-block-on.rs:10:1
   |
10 | #[rye::test_case(block_on = "custom_block_on")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
   | |                                             |
   | |                                             `#[rye::test_case(block_on = "custom_block_on")]` is later dropped here
   | await occurs here, with `#[rye::test_case(block_on = "custom_block_on")]` maybe used later
   | has type `rye::test_case::TestCase`

error: future cannot be sent between threads safely
  --> $DIR/02-custom-block-on.rs:10:29
   |
3  | fn custom_block_on<F>(fut: F) -> F::Output
   |    ---------------
4  | where
5  |     F: std::future::Future + Send + 'static,
   |                              ---- required by this bound in `custom_block_on`
...
10 | #[rye::test_case(block_on = "custom_block_on")]
   |                             ^^^^^^^^^^^^^^^^^ future returned by `nested` is not `Send`
   |
   = help: within `rye::test_case::TestCase`, the trait `std::marker::Sync` is not implemented for `std::rc::Rc<std::cell::RefCell<std::collections::HashMap<&'static rye::section::SectionId, rye::section::SectionData>>>`
note: future is not `Send` as this value is used across an await
  --> $DIR/02-custom-block-on.rs:10:1
   |
10 | #[rye::test_case(block_on = "custom_block_on")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-
   | |                                             |
   | |                                             `#[rye::test_case(block_on = "custom_block_on")]` is later dropped here
   | await occurs here, with `#[rye::test_case(block_on = "custom_block_on")]` maybe used later
   | has type `&rye::test_case::TestCase`

error: future cannot be sent between threads safely
  --> $DIR/02-custom-block-on.rs:10:29
   |
3  | fn custom_block_on<F>(fut: F) -> F::Output
   |    ---------------
4  | where
5  |     F: std::future::Future + Send + 'static,
   |                              ---- required by this bound in `custom_block_on`
...
10 | #[rye::test_case(block_on = "custom_block_on")]
   |                             ^^^^^^^^^^^^^^^^^ future returned by `nested` is not `Send`
   |
   = help: within `impl std::future::Future`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<()>`
note: future is not `Send` as this value is used across an await
  --> $DIR/02-custom-block-on.rs:23:9
   |
15 |     let rc = std::rc::Rc::new(());
   |         -- has type `std::rc::Rc<()>`
...
23 |         async {}.pending_once().await;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ await occurs here, with `rc` maybe used later
...
37 | }
   | - `rc` is later dropped here
